dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyPath.java

```diff
@@ -100,16 +107,66 @@ public int size() {
   }
 
   public Artifact getLeaf() {
-    return path.get(size() - 1);
+    return path.get(size() - 1).getArtifact();
   }
 
-  public List<Artifact> getPath() {
-    return path;
+  public ImmutableList<Artifact> getPath() {
+    return path.stream().map(Node::getArtifact).collect(toImmutableList());
   }
 
   // TODO think about index out of bounds
   public Artifact get(int i) {
-    return path.get(i);
+    return path.get(i).getArtifact();
+  }
+
+  /** Node in a dependency tree, holding scope and optional flag. */
+  private static class Node {
+    private Artifact artifact;
+    private String scope;
+    private Boolean optional;
+
+    private Node(Artifact artifact, String scope, boolean optional) {
+      this.artifact = checkNotNull(artifact);
+      this.scope = checkNotNull(scope);
+      this.optional = optional;
+    }
+
+    private Artifact getArtifact() {
+      return artifact;
+    }
+
+    private String getScope() {
+      return scope;
+    }
+
+    private Boolean isOptional() {
+      return optional;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (this == other) {
+        return true;
+      }
+      if (!(other instanceof Node)) {
+        return false;
+      }
+      Node that = (Node) other;
+      return optional == that.optional
+          && artifact.equals(that.artifact)
+          && scope.equals(that.scope);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(artifact, scope, optional);
+    }
+
+    @Override
+    public String toString() {
+      return String.format(
```

Comment:
This opaque. Breaking it up into steps with local variables and perhaps using concatenation instead of format should make it clearer.